Remake the 2d platformer demo

NathanLovato opened the pull request:
This Godot project is a complete rewrite of the official Platformer 2D demo. This is part of a proposal to improve and harmonize the official demo projects. We want to:

- Teach good programming practices with Godot.
- Bring demos to a higher quality standard.
- Keep unifying the code in the demos, a big task that @aaronfranke has been working on lately.

See the [open issue](https://github.com/godotengine/godot-demo-projects/issues/390) for more information.

This new demo showcases features from the original and works with Godot 3.2.
This is the same demo from https://github.com/GDQuest/godot-demo-projects-remake, coded by @johnnygossdev, marked as a co-author of the commit, first [reviewed by me](https://github.com/GDQuest/godot-kickstarter-2019/pull/58), and then by @aaronfranke.

We hope you'll like it!

NathanLovato commented:
I can write a README for the demo if you'd like, I just need to know if there's a format or anything, in particular, that you need there. The source code contains quite a lot of comments for learning purposes.

aaronfranke commented on 2d/platformer/src/UserInterface/PauseMenu.gd:
```diff
@@ -0,0 +1,26 @@
+	visible = false
+
+
+func _on_ResumeButton_pressed() -> void:
```
Akien said he doesn't want static typing used in most of the demo projects. I personally like static typing and I'm not against using it in some places if it improves readability. However, `-> void` is one of the least useful things in static typing, and it doesn't seem necessary here, so I don't think this should be here.

NathanLovato replied:
I had removed the type hints, but it's probably because I forgot to remove the static type hints when switching to the project at some point. I've got a script to remove them all.

In the previous discussion about code order and the style guide, I think we chose to use dynamic only unless the project or code snippets showcase static typing. And it's in the docs iirc. So I'll just remove them all.

NathanLovato commented:
Thanks, somehow I didn't get or missed the notification! I'll address the requested changes ASAP.

NathanLovato commented:
I've just addressed everything except for the background for now. I'm trying to get the game window to stay within supported screen ratios:

- I'd rather not have black bars on the game window
- You don't want the user to be able to resize to any window sizes with the expand stretch mode as they can see a lot more of the level, thus break the game/level design.

Do you know how to do that well? I have some code that does keep the game within size ratios, but there are bugs as I'm forcing a window resize every time the viewport resizes.

Here's the current code that's failing:

```gdscript
# Only allow window size ratios between 4:3 and 2:1, which are common display ratios
export var screen_ratio_range = Vector2(4.0 / 3.0, 2.0 / 1.0)

[...]

func _on_SceneTree_screen_resized():
	var window_size = OS.get_window_size()
	var ratio = window_size.x / window_size.y
	var ratio_clamped = clamp(ratio, screen_ratio_range.x, screen_ratio_range.y)
	# Use the is_equal_approx function to compare two decimal values, as computers 
	# don't store floats with absolute accuracy.
	var is_ratio_in_range = is_equal_approx(ratio_clamped, ratio)
	
	if not is_ratio_in_range:
		OS.window_size = (window_size / ratio * window_size.x).round()
```

aaronfranke commented on 2d/platformer/README.md:
```diff
@@ -0,0 +1,25 @@
+# 2D Platformer #
```
It's interesting that this works (using # at the end, TIL), but it's not common practice in Markdown, so I'd suggest avoiding it. Should look like:

    # 2D Platformer

NathanLovato replied:
This works everywhere, that's how emacs inserts markdown headings. But I'll remove it if it's a problem.

aaronfranke commented on 2d/platformer/README.md:
```diff
@@ -0,0 +1,25 @@
+
+It shows you how to code characters and physics-based objects in a real game context. This is a relatively complete demo where the player can jump, walk on slopes, fire bullets, interact with enemies, and more. It contains one closed level, and the player is invincible, unlike the enemies.
+
+You will find most of the demo’s content in the `Level.tscn`. You can open it from the default `Game.tscn` scene, or double click on `Level.tscn` in the `src/Level/` directory.
```
"in the `Level.tscn` [scene/file]", or maybe drop the article ("the") if you think it's too many words: "in `Level.tscn`"

NathanLovato commented:
I copied the settings from the original demo. So now there are black bars at the top and the bottom of the window when you make it taller than the base ratio. The last commit also prevents the background from blinking, and I extended it so you don't see the clear color when the background is wide.

I also tweaked the markdown.

aaronfranke commented on 2d/platformer/src/Main/Game.gd:
```diff
@@ -0,0 +1,31 @@
+onready var _pause_menu = $InterfaceLayer/PauseMenu
+
+
+func _init() -> void:
```
🤔

johnnygossdev replied:
Hi! I'm just jumping in. What's the issue here? Is it the use of `-> void:`?

I've just tested and the platforms in the background are visible for me but the clouds are indeed to the right. This should be a simple fix of offsetting the `ParallaxBackground`.

I've made some tweaks and I can't find issue with wide resolutions.
![platformer-wide](https://user-images.githubusercontent.com/3285214/74335583-6dbfb400-4d94-11ea-95f7-826b750eb75c.png)

I can make other touch ups and commit if that's all good @NathanLovato

NathanLovato replied:
Yeah sure, by the way @aaronfranke you have write access to my branch and PR, so when you find a typo or something, you can remove it right away - it'll take you less time than asking us to remove one.

I'm jumping between many projects every day, all but this one use type hints and the related completion. I don't even notice them to be honest. I'll check if I can't add a local pre-commit commit hook to run the script that cleans type hints for future contributions. Or at least I should be able to do that on save in Emacs.

NathanLovato replied:
Also, I can confirm the platforms do show up here - and yes the clouds are offset, I'll move them to the left. I'll also try to see what could make the platforms disappear.

![Screenshot from 2020-02-12 07-40-31](https://user-images.githubusercontent.com/12694995/74340167-0b9e8900-4d6b-11ea-94b5-5ce37b826f1b.png)

NathanLovato replied:
Is there a way to change editor settings per-project? Type hints seems like something that could be overriden per file or per project.

Calinou replied:
> Is there a way to change editor settings per-project? Type hints seems like something that could be overriden per file or per project.

Not yet, although I suppose we could move the type hint setting from the Editor Settings to the Project Settings. If we really need it, we could keep the editor setting and make it an enum with 3 values: "Project Default", "Always", "Never".

NathanLovato commented:
Here, like yesterday, the platforms show at every window size. I fixed the clouds, but also tried to change the parallax layer setup so they don't get culled - they had their origin in the top-left of the game window, maybe this can create some edge cases?

I also removed the type hint. 

![Screenshot from 2020-02-12 07-50-38](https://user-images.githubusercontent.com/12694995/74340865-608ecf00-4d6c-11ea-8510-04e448abca90.png)

@johnnygossdev I took care of this one because you don't have write access to my branch and it just took a few minutes. It's a matter of adding the final touches at this point.

johnnygossdev commented:
Sure no problem! :+1:

NathanLovato commented:
Just added some clouds for when making the window too long

![Screenshot from 2020-02-12 07-54-17](https://user-images.githubusercontent.com/12694995/74341196-e579e880-4d6c-11ea-8adf-a444863bdc79.png)

NathanLovato commented:
Ah, I had forgotten to change the code order, thanks for cleaning this up!

NathanLovato commented:
> Once concern is that it might be over-documented in a few places, but I think that's fine since I would expect this demo to be viewed by beginners who need this information.

Yes, so that was the idea. We assumed that this would be a go-to demo for beginners as it's the kind of thing a newcomer may want to do with Godot.

Also, we've designed it with the idea of not writing and having to maintain a long text-based, step-by-step tutorial, which is difficult to keep up to date. So all the explanations are inline.

Thanks for taking the time to review and make fixes!

Sslaxx commented on 2d/platformer/src/UserInterface/PauseMenu.gd:
```diff
@@ -0,0 +1,26 @@
+extends Control
```
Seems odd that this file is entirely uncommented.

NathanLovato replied:
It's not the core of the demo, that's about 2D platforming. I'd recommend following a tutorial dedicated to pausing, as half of the work is happening in the inspector, with the pause mode on the nodes.

Sslaxx replied:
I meant within the context of the demos, not tutorials or other documentation. If the actual pause demo is going to be updated as well, at least giving a pointer to it might here be useful.

In general, if something is going to be demo-ed somewhere else in more detail, pointing to that in another demo might be helpful.

NathanLovato replied:
Sorry, I may not get what you mean. You mean we could add a comment telling people to go look at a dedicated demo about pausing? If so sure, anyone can commit something like that anytime.

Sslaxx replied:
Telling people which other demo would demonstrate a feature in more detail, yes.
